### DISCUSSION TITLE
Help with loader context type safety.

### DISCUSSION THREAD

**Discussion by dosmond** [TanStack/router#2668] [Q&A] [answered]
Hi there, 

I'm trying to follow the [Kitchen Sink with React Query File Based](https://github.com/TanStack/router/tree/main/examples/react/kitchen-sink-react-query-file-based) example and I'm unsure if I'm doing something wrong.

I start by setting up the query client and the router, including the `queryClient` in the router context and setting the `defaultPreloadStaleTime` to 0 because I'm using react query for the caching

```typescript
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      retry: false,
    },
  },
});

const router = createRouter({
  routeTree,
  defaultPreload: "intent",
  context: { queryClient },
  defaultPreloadStaleTime: 0,
});

declare module "@tanstack/react-router" {
  interface Register {
    router: typeof router;
  }
}
```

When trying to access the context in the loader I get the following typescript error
![image](https://github.com/user-attachments/assets/13126fcf-640a-47de-8d9b-ef091a05268b)

In order to fix this error I need to explicitly define the type of the context in the loader fn like this

```typescript
loader: (opts) =>
    (opts.context as { queryClient: QueryClient }).queryClient.ensureQueryData(
      categoryQueryOptions
    ),
```
Is this expected?

**Comment by thomasballinger** [comment, marked as answer]
@dosmond did you perhaps forget this bit, https://github.com/TanStack/router/blob/main/examples/react/kitchen-sink-react-query-file-based/src/routes/__root.tsx#L5, using createRootRouterWithContext?

**Reply by thomasballinger** [reply]
https://github.com/TanStack/router/blob/1a55b977664cf76e821f41927418d3eb3903d769/examples/react/kitchen-sink-react-query-file-based/src/routes/__root.tsx#L19-L24

**Reply by dosmond** [reply]
Yes that was it! Thank you so much!
